File: frontend/src/types/metric.ts
export interface Metric {
  id?: string;
  job_id: string;
  epoch: number;
  step: number;
  loss: number;
  accuracy?: number;
  learning_rate?: number;
  throughput?: number;
  gpu_utilization?: number;
  memory_used_gb?: number;
  custom_metrics?: Record<string, any>;
  timestamp: string;
}

export interface MetricSummary {
  total_points: number;
  returned_points: number;
  best_loss: number;
  best_accuracy: number;
  current_loss: number;
  current_accuracy: number;
}

export interface MetricsResponse {
  job_id: string;
  metrics: Metric[];
  summary: MetricSummary;
}

export interface MetricUpdate {
  type: 'metric_update' | 'job_complete';
  job_id: string;
  epoch: number;
  step: number;
  metrics: {
    loss: number;
    accuracy?: number;
    learning_rate?: number;
    throughput?: number;
  };
  timestamp: string;
  status?: string;
  final_metrics?: {
    loss: number;
    accuracy: number;
  };
  completed_at?: string;
}

File: backend/app/schemas/metric.py
from pydantic import BaseModel
from typing import Optional, Dict, Any, List
from datetime import datetime


class MetricResponse(BaseModel):
    id: Optional[str] = None
    job_id: str
    epoch: int
    step: int
    loss: float
    accuracy: Optional[float] = None
    learning_rate: Optional[float] = None
    throughput: Optional[float] = None
    gpu_utilization: Optional[float] = None
    memory_used_gb: Optional[float] = None
    custom_metrics: Optional[Dict[str, Any]] = None
    timestamp: datetime

    class Config:
        from_attributes = True


class MetricSummary(BaseModel):
    total_points: int
    returned_points: int
    best_loss: float
    best_accuracy: float
    current_loss: float
    current_accuracy: float


class MetricsResponse(BaseModel):
    job_id: str
    metrics: List[MetricResponse]
    summary: MetricSummary


class MetricUpdate(BaseModel):
    type: str  # metric_update or job_complete
    job_id: str
    epoch: int
    step: int
    metrics: Dict[str, float]
    timestamp: datetime
    status: Optional[str] = None
    final_metrics: Optional[Dict[str, float]] = None
    completed_at: Optional[datetime] = None
